[FedEx Integration]

patilsangram opened the pull request:
This is FedEx Integration PR covering,

1. Required Masters/DocType and updates.
2. Re-thought/structure of Shipping Plan as a standalone DocType
3. Association with DN as reference DocType (Currently supported)
4. Provision to Make Shipping Plan from DN.
4. Validations on Shipping Plan.

Please, do review and comment on the same.

Expected next level of updates will include,

1. Revisit field mapping and updates  
2. FedEx related validations
3. Filters as appropriate.
4. Hit FedEx 

References..
[OLD PR](https://github.com/frappe/erpnext/pull/11541)
[Issue](https://github.com/frappe/erpnext/issues/11468)
[Discuss Thread](https://discuss.erpnext.com/t/erpnext-foundation-pune-chapter-code-sprint-3-fedex-integration/32803)

Indictrans.

patilsangram committed:
[Shipping plan with DN ref and Shipping Plan Item, Package tables added on shipping Plan]

patilsangram committed:
[field level changes on shipping plan]

patilsangram committed:
[added Master's - FedEx Postal Code, Package Item]

patilsangram committed:
[Shipping Plan validations and filters]

patilsangram committed:
[Make Shipping Plan provision from DN]

patilsangram committed:
[fix] codecy issue

revant commented:
- Can Postal Code be generalised instead of Fedex Postal Code? It can be re-used
- `Package` to `Fedex Package` can be dynamic link instead of Link? DHL Package/UPS package can be set like that?
- Shipper can become linked with supplier? Supplier contacts can be used for multiple persons interacting from shipper party
- Shipping Plan tied with only delivery note?
    - Can it be also tied with Purchase Receipt?
    - Stock Entry -> Material Issue / Receipt can also generate shipping in some cases. Intra warehouse shipping.

revant reviewed (commented):
Refer Comment

patilsangram commented:
Thank you @revant. I will do the necessary changes and send it again.

saurabh6790 commented:
@patilsangram Also add some screenshots

patilsangram commented:
- Can Postal Code be generalised instead of Fedex Postal Code? It can be re-used
 
>> Okay we can rename it as Postal Code

- Package to Fedex Package can be dynamic link instead of Link? DHL Package/UPS package can be set like that?
	
>> If we create Package no as Dynamic field then we have to create Package Type as Link field.

- Shipper can become linked with supplier? Supplier contacts can be used for multiple persons interacting from shipper party
	
>> We can add Supplier link on shipper

- Shipping Plan tied with only delivery note? Can it be also tied with Purchase Receipt?
Stock Entry -> Material Issue / Receipt can also generate shipping in some cases. Intra company warehouse shipping.
	
>> Yes, it is in plan.Currently we are mapped it from Delivery Note. In future we can provide Make option on above mentioned form to Make Shipping Plan from it.


- Edit: Also I Feel, Company needs child table for various different numbers like TIN, PAN, etc. because we need to keep adding removing such numbers for domains, integrations, localisations, etc.

>> Let's wait for more comments from community on this.

revant commented:
> If we create Package no as Dynamic field then we have to create Package Type as Link field.

For this PR we can go with Link and have a complete integration with FedEx first.

> Let's wait for more comments from community on this.

:+1: We'll wait for comments.

patilsangram committed:
[renamed fedex postal code to postal code, added supplier link to shipper]

patilsangram commented:
Hi,

As per suggestions, I have renamed FedEx Postal Code to Postal Code. 
Also, added supplier link on Shipper Master. Let me know if anything. So, we can move ahead.

patilsangram commented:
@saurabh6790 

These are the some Master's screenshot.

- FedEx Settings

![image](https://user-images.githubusercontent.com/14306547/35900762-a27491ea-0bf9-11e8-95f6-2cb3fce41d01.png)

- Shipper

![image](https://user-images.githubusercontent.com/14306547/35900668-2ecf8f7e-0bf9-11e8-8e8f-b0e84e771a75.png)

- Shipping Plan

![image](https://user-images.githubusercontent.com/14306547/35901075-fb8e7650-0bfa-11e8-99ae-854912b625ef.png)

Shipping Plan Item [Table] - Item details
![image](https://user-images.githubusercontent.com/14306547/35901069-f17a77ea-0bfa-11e8-9375-949f4278ce6e.png)

Package [Table] - for Package details
![image](https://user-images.githubusercontent.com/14306547/35901123-3a515ed4-0bfb-11e8-89ef-30feb7519869.png)

Package Item [Table] - for item distribution in Packges
![image](https://user-images.githubusercontent.com/14306547/35901181-654a9a7e-0bfb-11e8-845c-bc92239dc896.png)

[ Other details ]
![image](https://user-images.githubusercontent.com/14306547/35900938-7486b320-0bfa-11e8-8bea-2725fd305d25.png)

- Postal Code

![image](https://user-images.githubusercontent.com/14306547/35901425-86a84d82-0bfc-11e8-8606-7a7e139153d5.png)

patilsangram committed:
[fix] resolved conflicts

patilsangram committed:
[fix] resolved requirement conflicts

patilsangram committed:
[Shipping Plan] filters/queries, total handling unit calculation

patilsangram committed:
[Shipping Plan validations] validate - postal_code,package count, package details, set net weight

patilsangram committed:
[missing shipping plan fields]

patilsangram committed:
[fix] invalide method name

patilsangram committed:
[made shipping plan submittable, added picup fields]

patilsangram committed:
[Schedule Pickup API]

patilsangram committed:
[fix] import fail

patilsangram committed:
[fix] codacy issue

joshreeder commented:
What needs to happen on this?

patilsangram commented:
We would require the FedEx credentials to validate the API's. We expired the earlier one available and not sure how much time will it take to get one fresh. If anybody has an access can spare for testing would be really great to expedite on hold implementation. We are sure that will not charge you anything since we can cancel a planned pick-up (or generally not charged unless pick-up from your end).

rohitwaghchaure commented on erpnext/logistics/controller/fedex_controller.py:
```diff
@@ -0,0 +1,66 @@
+		which handles API like Shipment, Tracking, GET Rate
+		& other supplementary tasks. """
+
+	uom_mapper = {"Kg":"KG", "LB":"LB"}
```
Why uom mapper?
